core/src/main/java/org/elasticsearch/script/TemplateService.java

```diff
@@ -0,0 +1,239 @@
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonMap;
+
+/**
+ * Source of templates (think Mustache or StringTemplate).
+ */
+public class TemplateService implements ClusterStateListener {
+    public interface Backend extends ScriptEngineService {} // TODO customize this for templates
+
+    private static final String DEFAULT_CONTENT_TYPE = "text/plain";
+    private static final Logger logger = ESLoggerFactory.getLogger(TemplateService.class);
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(logger);
+    private final Backend backend;
+    private final ScriptPermits scriptPermits;
+    private final CachingCompiler<CacheKey> compiler;
+
+    public TemplateService(Settings settings, Environment env,
+            ResourceWatcherService resourceWatcherService, Backend backend,
+            ScriptContextRegistry scriptContextRegistry, ScriptSettings scriptSettings,
+            ScriptMetrics scriptMetrics) throws IOException {
+        Objects.requireNonNull(scriptContextRegistry);
+
+        this.backend = backend;
+        this.scriptPermits = new ScriptPermits(settings, scriptSettings, scriptContextRegistry);
+        this.compiler = new CachingCompiler<CacheKey>(settings, env, resourceWatcherService,
+                scriptMetrics, "template") {
+            @Override
+            protected CacheKey cacheKeyForFile(String baseName, String extension) {
+                if (false == backend.getType().equals(extension)) {
+                    /* For backwards compatibility templates are in the scripts directory and we
+                     * must ignore everything but templates. */
+                    return null;
+                }
+                return new CacheKey(baseName, DEFAULT_CONTENT_TYPE);
+            }
+
+            @Override
+            protected CacheKey cacheKeyFromClusterState(StoredScriptSource scriptMetadata) {
+                String contentType = DEFAULT_CONTENT_TYPE;
+                if (scriptMetadata.getOptions() != null) {
+                    contentType = scriptMetadata.getOptions().get(Script.CONTENT_TYPE_OPTION);
+                    if (contentType == null) {
+                        contentType = DEFAULT_CONTENT_TYPE;
+                    }
+                }
+                return new CacheKey(scriptMetadata.getCode(), contentType);
+            }
+
+            @Override
+            protected StoredScriptSource lookupStoredScript(ClusterState clusterState,
+                    CacheKey cacheKey) {
+                if (clusterState == null) {
+                    return null;
+                }
+                MetaData metaData = clusterState.metaData();
+                if (metaData == null) {
+                    return null;
+                }
+                ScriptMetaData scriptMetaData = clusterState.metaData().custom(ScriptMetaData.TYPE);
+                if (scriptMetaData == null) {
```

Comment:
I certainly *can* use optional and the map for these. If you'd prefer I'm happy to do it.